File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2020/Y20D17.kt
package org.gristle.pdxpuzzles.advent.y2020

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.objects.MCoord
import org.gristle.pdxpuzzles.utilities.objects.toGrid

class Y20D17(input: String) : Day {
    private val startGrid = input.toGrid()

    private fun getCubes(dimensions: Int): Int {
        var space = mutableSetOf<MCoord>()

        startGrid.mapIndexedNotNull { index, c ->
            if (c == '.') null else index
        }.forEach { index ->
            val gridCoord = startGrid.coordOf(index)
            val coordinates = (0 until dimensions).map { dim ->
                when (dim) {
                    0 -> gridCoord.x
                    1 -> startGrid.coordOf(index).y
                    else -> 0
                }
            }
            space.add(MCoord(coordinates))
        }

        val bounds = (0 until dimensions).map { dim ->
            when (dim) {
                0 -> startGrid.xIndices
                1 -> startGrid.yIndices
                else -> 0..0
            }
        }.toMutableList()

        for (i in 1..6) {
            val newSpace = mutableSetOf<MCoord>()
            (0 until dimensions).forEach { dim ->
                val bound = bounds[dim]
                bounds[dim] = (bound.first - 1)..(bound.last + 1)
            }

            val coordinates = bounds.drop(1).fold(bounds[0].toList().map { listOf(it) }) { acc, intRange ->
                acc.flatMap { intList -> intRange.map { intList + it } } // generate lists of list
            }

            coordinates.forEach { coordinate ->
                val mc = MCoord(coordinate)
                val neighbors = mc.getNeighbors().count { space.contains(it) }
                if (neighbors == 3 || (neighbors == 2 && space.contains(mc))) newSpace.add(mc)
            }

            space = newSpace
        }

        return space.size
    }

    override fun part1() = getCubes(3)

    override fun part2() = getCubes(4)
}

fun main() = Day.runDay(Y20D17::class)

//    Class creation: 20ms
//    Part 1: 346 (83ms)
//    Part 2: 1632 (1244ms)
//    Total time: 1348ms

File: rust/advent/src/bin/2020/y20d17.rs
use std::collections::HashSet;
use std::ops::Range;
use advent::utilities::get_input::get_input;
use utilities::structs::coord::Coord;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};
use utilities::structs::str_grid::StrGrid;

type Input<'a> = StrGrid<'a>;
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(20, 17).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &'_ str) -> Input<'_> {
    StrGrid::new(input).unwrap()
}

fn get_cubes<const DIMENSIONS: usize>(start: &StrGrid) -> Output {
    let mut space: HashSet<Coord<i64, DIMENSIONS>> = start.s.iter().enumerate()
        .filter(|&(_, &b)| b != b'.' && b != b'\n')
        .map(|(idx, _)| {
            let pos = start.idx_to_coord(&idx);
            let coordinates: [i64; DIMENSIONS] = std::array::from_fn(|dimension| {
                match dimension {
                    0 => pos.x() as i64,
                    1 => pos.y() as i64,
                    _ => 0,
                }
            });
            Coord::new(coordinates)
        })
        .collect();

    let mut bounds: [Range<i64>; DIMENSIONS] = std::array::from_fn(|dimension| {
        match dimension {
            0 => 0..start.width as i64 - 1,
            1 => 0..start.height as i64,
            _ => 0..1,
        }
    });

    for _ in 0..6 {
        let mut new_space: HashSet<Coord<i64, DIMENSIONS>> = HashSet::new();

        // expand bounds by one in each direction in each dimension
        for dimension in 0..DIMENSIONS {
            let bound = &bounds[dimension];
            bounds[dimension] = Range {start: bound.start - 1, end: bound.end + 1 };
        }

        let mut bounds_iter = bounds.iter();
        let init: Vec<Vec<i64>> = bounds_iter
            .next()
            .map(|rng| rng.clone().map(|it| vec![it]).collect())
            .unwrap();
        let coordinates = bounds_iter.fold(init, |acc, rng| {
            acc.into_iter()
                .flat_map(|int_list| {
                    rng.clone().map(move |new| {
                        int_list.iter().cloned().chain(std::iter::once(new)).collect()
                    })
                })
                .collect()
        });
        for coordinate in coordinates {
            let pos: Coord<i64, DIMENSIONS> = Coord::new(coordinate.try_into().unwrap());
            let neighbors = pos.get_neighbors()
                .filter(|neighbor| space.contains(neighbor))
                .count();
            if neighbors == 3 || (neighbors == 2 && space.contains(&pos)) {
                new_space.insert(pos);
            }
        }
        space = new_space;
    }
    space.len()
}

fn part1(start: &Input) -> Output {
    get_cubes::<3>(start)
}

fn part2(start: &Input) -> Output {
    get_cubes::<4>(start)
}

#[test]
fn default() {
    let input = get_input(20, 17).unwrap();
    let input = parse_input(&input);
    assert_eq!(346, part1(&input));
    assert_eq!(1632, part2(&input));
}

// Input parsed (30μs)
// 1. 346 (11ms)
// 2. 1632 (356ms)
// Total: 367ms